jsondiff.go

```diff
@@ -0,0 +1,886 @@
+			// Recursively compare and colorize slices
+			expectedText, actualText := compareAndColorizeSlices(v1, v2, indent+"  ", red, green)
+			expect.WriteString(fmt.Sprintf("%s\"%s\": [\n%s\n%s]\n", indent, key, expectedText, indent))
+			actual.WriteString(fmt.Sprintf("%s\"%s\": [\n%s\n%s]\n", indent, key, actualText, indent))
+			return
+		}
+		// If types do not match, write the key-value pairs with colors
+		writeKeyValuePair(expect, key, val1, indent, red)
+		writeKeyValuePair(actual, key, val2, indent, green)
+
+	// Default case for other types
+	default:
+		// Check if the values are not deeply equal
+		if !reflect.DeepEqual(val1, val2) {
+			// Marshal values to pretty-printed JSON strings
+			val1Str, err := json.MarshalIndent(val1, "", "  ")
+			if err != nil {
+				return
+			}
+			val2Str, err := json.MarshalIndent(val2, "", "  ")
+			if err != nil {
+				return
+			}
+			// Colorize the differences in the values
+			c := color.FgRed
+			offsetsStr1, offsetsStr2, _ := diffArrayRange(string(val1Str), string(val2Str))
+			expectDiff := breakSliceWithColor(string(val1Str), &c, offsetsStr1)
+			c = color.FgGreen
+			actualDiff := breakSliceWithColor(string(val2Str), &c, offsetsStr2)
+			expect.WriteString(breakLines(fmt.Sprintf("%s\"%s\": %s,\n", indent, key, string(expectDiff))))
+			actual.WriteString(breakLines(fmt.Sprintf("%s\"%s\": %s,\n", indent, key, string(actualDiff))))
+			return
+		}
+		// If values are equal, write the value without color
+		valStr, err := json.MarshalIndent(val1, "", "  ")
+		if err != nil {
+			return
+		}
+		expect.WriteString(fmt.Sprintf("%s\"%s\": %s,\n", indent, key, string(valStr)))
+		actual.WriteString(fmt.Sprintf("%s\"%s\": %s,\n", indent, key, string(valStr)))
+
+	}
+}
+
+// separateAndColorize separates the diff string into expected and actual strings, applying color where appropriate.
+// diffStr: The input string representing the differences.
+// noise: A map containing noise elements to be ignored during processing.
+// Returns two strings: the colorized expected and actual differences.
+func separateAndColorize(diffStr string, noise map[string][]string) (string, string) {
+	lines := strings.Split(diffStr, "\n") // Split the diff string into lines.
+	lines = insertEmptyLines(lines)       // Insert empty lines between consecutive elements with the same symbol.
+	// Initialize maps and arrays to store the expected and actual values.
+	expectMap := make(map[string]interface{}, 0)
+	actualMap := make(map[string]interface{}, 0)
+	expectsArray := make([]interface{}, 0)
+	actualsArray := make([]interface{}, 0)
+	var isExpectMap, isActualMap bool
+	expect, actual := "", ""
+
+	// Iterate over the lines, processing each line and the next line together.
+	for i := 0; i < len(lines)-1; i++ {
+		var expectKey, actualKey string
+		line := lines[i]
+		nextLine := lines[i+1]
+
+		// Process lines that start with a '-' indicating expected differences.
+		if len(line) > 0 && line[0] == '-' && i != len(lines)-1 {
```

Comment:
yes ,I've explicitly appended empty string in the lines in 	

lines = insertEmptyLines(lines)       // Insert empty lines between consecutive elements with the same symbol.

So that two ++ signs don't come after each other